### PR TITLE
First implementation of Photon Conversion

### PR DESCRIPTION
**Pull request by amaringarcia**

### CODE REVIEW

**Comment by njacazio** [comment]
Hi @amaringarcia tell us when ready for review

**Comment by amaringarcia** [comment]
Everything compiles, DelphesO2 runs, the AODs are filled. The only point missing is to check that the resolution and the efficiency is what I have plug in. I am writing a program for that. Any macro/program that reads MC particle and compared to reco one will help. 
So, in my opinion the review can already happen.

File: examples/aod/createO2tables.C [hunk 1]
```diff
@@ -285,6 +298,23 @@ int createO2tables(const char* inputFile = "delphes.root",
         dNdEta += 1.f;
       }
       FillTree(kMcParticle);
+
```
Comment (njacazio) [inline examples/aod/createO2tables.C]: Maybe it is worth to create the TLorentzVector only once outside of the loop and use it many times in the loop

File: examples/aod/createO2tables.C [hunk 2]
```diff
@@ -486,6 +516,16 @@ int createO2tables(const char* inputFile = "delphes.root",
         ftof_tracks_indices.push_back(std::pair<int, int>{ientry + eventOffset, fTrackCounter});
       }
 
+
+      // check if has Preshower
+      if(pre_shower.hasPreShower(*track)){
+	pres.fIndexCollisions = ientry + eventOffset;
```
Comment (njacazio) [inline examples/aod/createO2tables.C]: Could you use clang-format on the file? i.e. `clang-format -i createO2tables.C`

File: examples/aod/createO2tables.h [hunk 3]
```diff
@@ -56,6 +58,7 @@ const TString TreeName[kTrees] = {"O2collision",
                                   "O2v0",
                                   "O2cascade",
                                   "O2tof",
+				  "O2photonconv",
```
Comment (njacazio) [inline examples/aod/createO2tables.h]: Could you also run the formatter on the `.h` file?

File: rpythia8/rpythia8-box.cc [hunk 4]
```diff
@@ -73,16 +73,18 @@ int main(int argc, char** argv)
...
     std::cout << "Error: invalid PDG code \"" << pdg << "\" is not in the particle list" << std::endl;
     return 1;
   }
   if (!pythia.particleData.isLepton(pdg) &&
       !pythia.particleData.isHadron(pdg) &&
       !pythia.particleData.isResonance(pdg)) {
-    if (abs(pdg) < 1000000000) {
-      std::cout << "Error: invalid PDG code \"" << pdg << "\"" << std::endl;
-      return 1;
+    if (  abs(pdg) < 1000000000 && pdg != 22) {
```
Comment (njacazio) [inline rpythia8/rpythia8-box.cc]: Also this should be formatted with clang-format

File: src/PhotonConversion.cc [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+
+    }
+
+    /*****************************************************************/
+
+    bool PhotonConversion::hasPhotonConversion(const GenParticle& particle) const
+    {
+ 
+
```
Comment (njacazio) [inline src/PhotonConversion.cc]: It would be useful for the future to format also this piece of code

File: src/PhotonConversion.cc [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	return (gRandom->Uniform() < (convProb*eff) );
+      }
+      else {
+         const Float_t misConvProb = 0.0;
+         return (gRandom->Uniform() < misConvProb);
+      }
+     // AM plug-in the conversion probability * photon reconstruction efficiency
+
+      // auto x = track.XOuter * 0.1; // [cm]
+      // auto y = track.YOuter * 0.1; // [cm]
```
Comment (njacazio) [inline src/PhotonConversion.cc]: Could this be cleaned?

File: src/PhotonConversion.cc [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    bool PhotonConversion::makeSignal(const GenParticle& particle, TLorentzVector &photonConv)
+    {
+      const int pid = particle.PID;
+      if (pid != 22) {
+	return false;
+      }
+
+      // TLorentzVector p4True = particle.P4(); // true 4-momentum
+      // Float_t vtX = particle.X; // particle vertex X
+      // Float_t vtY = particle.Y; // particle vertex Y
```
Comment (njacazio) [inline src/PhotonConversion.cc]: Could this be cleaned as well?

File: src/PhotonConversion.cc [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      pSmeared.SetXYZM(pxSmeared,pySmeared,pzSmeared,0.);
+      return pSmeared;
+    }
+    /*****************************************************************/
+    // Double_t PhotonConversion::sigmaX(const Double_t eTrue)
+    // {
+    //   // Calculate sigma of photon coordinate smearing [cm]
+    //   // E is the photon energy
+    //   Double_t dX = sqrt(mPositionResolutionA*mPositionResolutionA + mPositionResolutionB*mPositionResolutionB/eTrue);
+    //   return dX;
```
Comment (njacazio) [inline src/PhotonConversion.cc]: This might be removed?

File: src/PhotonConversion.hh [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace o2
+{
+  namespace delphes
+  {
+
+    class PhotonConversion
+    {
+
+    public:
```
Comment (njacazio) [inline src/PhotonConversion.hh]: Maybe we could format also this

**Review by njacazio — CHANGES_REQUESTED** [review]
Hi, good work! If you could take care of the suggestions we can merge this

**Comment by amaringarcia** [comment]
I have followed your suggestions....
Please, let me know if it is ok now, or I should do other changes

File: rpythia8/rpythia8-box.cc [hunk 10]
```diff
@@ -73,18 +73,20 @@ int main(int argc, char** argv)
   }
 
   // check valid pdg code
-  if (!pythia.particleData.isParticle(pdg)) {
+  if (!pythia.particleData.isParticle(pdg) ) {
```
Comment (njacazio) [inline rpythia8/rpythia8-box.cc]: This change probably can be avoided

File: rpythia8/rpythia8-box.cc [hunk 11]
```diff
@@ -73,18 +73,20 @@ int main(int argc, char** argv)
...
     std::cout << "Error: invalid PDG code \"" << pdg << "\" is not in the particle list" << std::endl;
     return 1;
   }
   if (!pythia.particleData.isLepton(pdg) &&
       !pythia.particleData.isHadron(pdg) &&
       !pythia.particleData.isResonance(pdg)) {
+
```
Comment (njacazio) [inline rpythia8/rpythia8-box.cc]: idem

File: rpythia8/rpythia8-box.cc [hunk 12]
```diff
@@ -73,18 +73,20 @@ int main(int argc, char** argv)
...
     if (abs(pdg) < 1000000000 && abs(pdg) != 22) {
       std::cout << "Error: invalid PDG code \"" << pdg << "\"" << std::endl;
       return 1;
     } else if (abs(pdg) == 22) {
       std::cout << "PDG code \"" << pdg << "\" stands for a gamma" << std::endl;
+
```
Comment (njacazio) [inline rpythia8/rpythia8-box.cc]: idem

File: rpythia8/rpythia8.cc [hunk 13]
```diff
@@ -15,14 +15,7 @@ int main(int argc, char** argv)
...
-      ("help", "Print help messages")
-      ("nevents,n"      , po::value<int>(&nevents)->default_value(10), "Number of events")
-      ("config,c"       , po::value<std::string>(&config), "Configuration file")
-      ("output,o"       , po::value<std::string>(&output)->default_value("pythia8.hepmc"), "Output HepMC file")
-      ("inject-config"  , po::value<std::string>(&inject_config), "Injected event configuration file")
-      ("inject-nevents" , po::value<int>(&inject_nevents)->default_value(1), "Number of events to inject")
-      ("seed"           , po::value<int>(&seed)->default_value(1), "initial seed");
+    desc.add_options()("help", "Print help messages")("nevents,n", po::value<int>(&nevents)->default_value(10), "Number of events")("config,c", po::value<std::string>(&config), "Configuration file")("output,o", po::value<std::string>(&output)->default_value("pythia8.hepmc"), "Output HepMC file")("inject-config", po::value<std::string>(&inject_config), "Injected event configuration file")("inject-nevents", po::value<int>(&inject_nevents)->default_value(1), "Number of events to inject")("seed", po::value<int>(&seed)->default_value(1), "initial seed");
```
Comment (njacazio) [inline rpythia8/rpythia8.cc]: This is not easy to read, it would be best to not do it

File: examples/aod/createO2tables.C [hunk 14]
```diff
@@ -184,9 +189,20 @@ int createO2tables(const char* inputFile = "delphes.root",
   forward_rich_detector.setType(o2::delphes::RICHdetector::kForward);
   forward_rich_detector.setRadiusIn(forward_rich_radius_in);
 
+
+  // Photon Conversion Method
+  o2::delphes::PhotonConversion photon_conversion;
+  //  photon_conversion.setup();
```
Comment (njacazio) [inline examples/aod/createO2tables.C]: This could be cleaned

File: examples/aod/createO2tables.h [hunk 15]
```diff
@@ -478,11 +500,31 @@ void MakeTreeO2ftof()
 }
 
 struct {
-  // ALICE3 ECAL data
+
```
Comment (njacazio) [inline examples/aod/createO2tables.h]: This could  stay the same, probably it can be fixed by running clang-format

**Review by njacazio — CHANGES_REQUESTED** [review]
Dear @amaringarcia great work! Please consider the last comments and then we can merge it, as a general comment please avoid changing the code that is not strictly needed as it makes review harder. Also, don't forget to run `clang-format` as it allows to avoid some not needed changes (except for the rpythia code where the configuration becomes hard to read)

File: rpythia8/rpythia8.cc (hunk 13 above)
Comment (amaringarcia) [reply rpythia8/rpythia8.cc]: I am not aware of having changed this file

**Comment by njacazio** [comment]
Ok, great, let me try this to verify that it compiles and we are set

**Comment by amaringarcia** [comment]
Thanks a lot. Sorry for all the extra work.!

**Pull request merged**